cli/external_workspaces.go

```diff
@@ -0,0 +1,326 @@
+package cli
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/xerrors"
+
+	"github.com/google/uuid"
+
+	"github.com/coder/coder/v2/cli/cliui"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/pretty"
+	"github.com/coder/serpent"
+)
+
+type externalAgent struct {
+	AgentName  string `json:"-"`
+	AuthType   string `json:"auth_type"`
+	AuthToken  string `json:"auth_token"`
+	InitScript string `json:"init_script"`
+}
+
+func (r *RootCmd) externalWorkspaces() *serpent.Command {
+	orgContext := NewOrganizationContext()
+
+	cmd := &serpent.Command{
+		Use:   "external-workspaces [subcommand]",
+		Short: "External workspace related commands",
+		Handler: func(inv *serpent.Invocation) error {
+			return inv.Command.HelpHandler(inv)
+		},
+		Children: []*serpent.Command{
+			r.externalWorkspaceCreate(),
+			r.externalWorkspaceAgentInstructions(),
+			r.externalWorkspaceList(),
+		},
+	}
+
+	orgContext.AttachOptions(cmd)
+	return cmd
+}
+
+// externalWorkspaceCreate extends `coder create` to create an external workspace.
+func (r *RootCmd) externalWorkspaceCreate() *serpent.Command {
+	var (
+		orgContext = NewOrganizationContext()
+		client     = new(codersdk.Client)
+	)
+
+	cmd := r.create()
+	cmd.Use = "create [workspace]"
+	cmd.Short = "Create a new external workspace"
+	cmd.Middleware = serpent.Chain(
+		cmd.Middleware,
+		r.InitClient(client),
+		serpent.RequireNArgs(1),
+	)
+
+	createHandler := cmd.Handler
+	cmd.Handler = func(inv *serpent.Invocation) error {
+		workspaceName := inv.Args[0]
+		templateVersion := inv.ParsedFlags().Lookup("template-version")
+		templateName := inv.ParsedFlags().Lookup("template")
+		if templateName == nil || templateName.Value.String() == "" {
+			return xerrors.Errorf("template name is required for external workspace creation. Use --template=<template_name>")
+		}
+
+		organization, err := orgContext.Selected(inv, client)
+		if err != nil {
+			return xerrors.Errorf("get current organization: %w", err)
+		}
+
+		template, err := client.TemplateByName(inv.Context(), organization.ID, templateName.Value.String())
+		if err != nil {
+			return xerrors.Errorf("get template by name: %w", err)
+		}
```

Comment:
A lot of these queries are going to be repeated by the original handler, could we change the `createHandler` function into a model where the queries get run once, then a "validate" function that gets swapped for `coder create` or `coder external-workspaces create` gets executed (and all the queried objects get passed in), then the workspace gets created.

```go
workspace, workspaceBuild, err := createHandler(ctx, inv, createOptions{
  beforeCreate: func(ctx context.Context, template codersdk.Template, ...) error {},
})
```